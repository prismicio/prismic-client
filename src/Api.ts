import Predicates from './Predicates';
import { RequestHandler, RequestCallback } from './request';
import { Document } from './documents';
import { ApiCache } from './cache';
import ResolvedApi, { ApiData } from './ResolvedApi';
import HttpClient from './HttpClient';

export interface ApiOptions {
  accessToken?: string;
  requestHandler?: RequestHandler;
  req?: any;
  apiCache?: ApiCache;
  apiDataTTL?: number;
  proxyAgent?: any;
}

export default class PrismicApi {
  url: string;
  options: ApiOptions;
  apiDataTTL: number;
  httpClient: HttpClient;

  constructor(url: string, options?: ApiOptions) {
    this.options = options || {};
    this.url = url;
    if (this.options.accessToken) {
      const accessTokenParam = `access_token=${this.options.accessToken}`;
      this.url += (url.indexOf('?') > -1 ? '&' : '?') + accessTokenParam;
    }
    this.apiDataTTL = this.options.apiDataTTL || 5;
    this.httpClient = new HttpClient(
      this.options.requestHandler,
      this.options.apiCache,
      this.options.proxyAgent,
    );
  }

  /**
   * Fetches data used to construct the api client, from cache if it's
   * present, otherwise from calling the prismic api endpoint (which is
   * then cached).
   */
  get(cb?: RequestCallback<ResolvedApi>): Promise<ResolvedApi> {
    return this.httpClient.cachedRequest<ApiData>(this.url, { ttl: this.apiDataTTL }).then((data) => {
      const resolvedApi = new ResolvedApi(data, this.httpClient, this.options);
      cb && cb(null, resolvedApi);
      return resolvedApi;
    }).catch((error) => {
      cb && cb(error);
      throw error;
    });
  }
}
