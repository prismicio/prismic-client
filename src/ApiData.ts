export interface QuickRoutes {
  url: string;
  lastUpdate: number;
  enabled: boolean;
}

export interface ApiData {
  refs: Ref[];
  bookmarks: { [key: string]: string };
  types: { [key: string]: string };
  tags: string[];
  forms: { [key: string]: Form };
  quickRoutes: QuickRoutes;
  experiments: any;
  oauth_initiate: string;
  oauth_token: string;
  version: string;
  licence: string;
}
