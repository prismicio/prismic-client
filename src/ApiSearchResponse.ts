import { Document } from "./documents";

export default interface ApiSearchResponse<T = any> {
  page: number;
  results_per_page: number;
  results_size: number;
  total_results_size: number;
  total_pages: number;
  next_page: string;
  prev_page: string;
  results: Document<T>[];
}
