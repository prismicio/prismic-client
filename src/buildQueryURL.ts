import { castArray } from "./lib/castArray";

import { ValueOf, Ordering, Route } from "./types";

/**
 * Parameters for the Prismic REST API V2.
 *
 * {@link https://prismic.io/docs/technologies/introduction-to-the-content-query-api}
 */
export interface QueryParams {
	/**
	 * The secure token for accessing the API (only needed if your repository is
	 * set to private).
	 *
	 * {@link https://user-guides.prismic.io/en/articles/1036153-generating-an-access-token}
	 */
	accessToken?: string;

	/**
	 * The `pageSize` parameter defines the maximum number of documents that the
	 * API will return for your query.
	 *
	 * {@link https://prismic.io/docs/technologies/search-parameters-reference-rest-api#pagesize}
	 */
	pageSize?: number;

	/**
	 * The `page` parameter defines the pagination for the result of your query.
	 *
	 * {@link https://prismic.io/docs/technologies/search-parameters-reference-rest-api#page}
	 */
	page?: number;

	/**
	 * The `after` parameter can be used along with the orderings option. It will
	 * remove all the documents except for those after the specified document in the list.
	 *
	 * {@link https://prismic.io/docs/technologies/search-parameters-reference-rest-api#after}
	 */
	after?: string;

	/**
	 * The `fetch` parameter is used to make queries faster by only retrieving the
	 * specified field(s).
	 *
	 * {@link https://prismic.io/docs/technologies/search-parameters-reference-rest-api#fetch}
	 */
	fetch?: string | string[];

	/**
	 * The `fetchLinks` parameter allows you to retrieve a specific content field
	 * from a linked document and add it to the document response object.
	 *
	 * {@link https://prismic.io/docs/technologies/search-parameters-reference-rest-api#fetchlinks}
	 */
	fetchLinks?: string | string[];

	/**
	 * The `graphQuery` parameter allows you to specify which fields to retrieve
	 * and what content to retrieve from Linked Documents / Content Relationships.
	 *
	 * {@link https://prismic.io/docs/technologies/graphquery-rest-api}
	 */
	graphQuery?: string;

	/**
	 * The `lang` option defines the language code for the results of your query.
	 *
	 * {@link https://prismic.io/docs/technologies/search-parameters-reference-rest-api#lang}
	 */
	lang?: string;

	/**
	 * The `orderings` parameter orders the results by the specified field(s). You
	 * can specify as many fields as you want.
	 *
	 * {@link https://prismic.io/docs/technologies/search-parameters-reference-rest-api#orderings}
	 */
	orderings?: Ordering | string | (Ordering | string)[];

	/**
	 * The `routes` option allows you to define how a document's `url` field is resolved.
	 *
	 * {@link https://prismic.io/docs/core-concepts/link-resolver-route-resolver#route-resolver}
	 */
	routes?: Route | string | (Route | string)[];

	/**
	 * The `brokenRoute` option allows you to define the route populated in the
	 * `url` property for broken Link or Content Relationship fields. A broken
	 * link is a Link or Content Relationship field whose linked document has been
	 * unpublished or deleted.
	 *
	 * {@link https://prismic.io/docs/core-concepts/link-resolver-route-resolver#route-resolver}
	 */
	brokenRoute?: string;
}

/**
 * Arguments for `buildQueryURL` to construct a Query URL.
 */
type BuildQueryURLParams = {
	/**
	 * Ref used to query documents.
	 *
	 * {@link https://prismic.io/docs/technologies/introduction-to-the-content-query-api#prismic-api-ref}
	 */
	ref: string;

	/**
	 * Ref used to populate Integration Fields with the latest content.
	 *
	 * {@link https://prismic.io/docs/core-concepts/integration-fields}
	 */
	integrationFieldsRef?: string;

	/**
	 * One or more predicates to filter documents for the query.
	 *
	 * {@link https://prismic.io/docs/technologies/query-predicates-reference-rest-api}
	 */
	predicates?: string | string[];
};

/**
 * Parameters in this map have been renamed from the official Prismic REST API
 * V2 specification for better developer ergonomics.
 *
 * These parameters are renamed to their mapped value.
 */
const RENAMED_PARAMS = {
	accessToken: "access_token",
} as const;

/**
 * A valid parameter name for the Prismic REST API V2.
 */
type ValidParamName =
	| Exclude<
			keyof QueryParams,
			keyof typeof RENAMED_PARAMS | keyof BuildQueryURLParams
	  >
	| ValueOf<typeof RENAMED_PARAMS>;

/**
 * Converts an Ordering to a string that is compatible with Prismic's REST API.
 * If the value provided is already a string, no conversion is performed.
 *
 * @param ordering - Ordering to convert.
 *
 * @returns String representation of the Ordering.
 */
const castOrderingToString = (ordering: Ordering | string): string =>
	typeof ordering === "string"
		? ordering
		: [
				ordering.field,
				ordering.direction === "desc" ? ordering.direction : undefined,
		  ]
				.filter(Boolean)
				.join(" ");

export type BuildQueryURLArgs = QueryParams & BuildQueryURLParams;

/**
 * Build a Prismic REST API V2 URL to request documents from a repository. The
 * paginated response for this URL includes documents matching the parameters.
 *
 * A ref is required to make a request. Request the `endpoint` URL to retrieve a
 * list of available refs.
 *
 * Type the JSON response with `Query`.
 *
 * {@link https://prismic.io/docs/technologies/introduction-to-the-content-query-api#prismic-api-ref}
 * {@link https://prismic.io/docs/technologies/query-predicates-reference-rest-api}
 *
 * @param endpoint - URL to the repository's REST API V2.
 * @param args - Arguments to filter and scope the query.
 *
 * @returns URL that can be used to request documents from the repository.
 */
export const buildQueryURL = (
	endpoint: string,
	args: BuildQueryURLArgs,
): string => {
	const { predicates, ...params } = args;

	const url = new URL(`documents/search`, `${endpoint}/`);

	if (predicates) {
		for (const predicate of castArray(predicates)) {
			url.searchParams.append("q", `[${predicate}]`);
		}
	}

	// Iterate over each parameter and add it to the URL. In some cases, the
	// parameter value needs to be transformed to fit the REST API.
	for (const k in params) {
		const name = (RENAMED_PARAMS[k as keyof typeof RENAMED_PARAMS] ??
			k) as ValidParamName;

		let value = params[k as keyof typeof params];

		if (name === "orderings") {
			const scopedValue = params[name];

			if (scopedValue != null) {
				const v = castArray(scopedValue)
					.map((ordering) => castOrderingToString(ordering))
					.join(",");

				value = `[${v}]`;
			}
		} else if (name === "routes") {
			if (typeof params[name] === "object") {
				value = JSON.stringify(castArray(params[name]));
			}
		}

		if (value != null) {
			url.searchParams.set(name, castArray(value).join(","));
		}
	}

	return url.toString();
};
