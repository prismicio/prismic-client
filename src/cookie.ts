/**
 * The well-known name of the cookie used to store a Prismic preview session's
 * ref.
 */
export const preview = "io.prismic.preview"
