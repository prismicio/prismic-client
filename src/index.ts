// Primary library API.
import { getRepositoryEndpoint } from "./getRepositoryEndpoint";
export { getRepositoryEndpoint };
/**
 * @deprecated Renamed to `getRepositoryEndpoint`.
 */
// TODO: Remove in v3.
export const getEndpoint = getRepositoryEndpoint;
export { getRepositoryName } from "./getRepositoryName";
export { getGraphQLEndpoint } from "./getGraphQLEndpoint";
export { isRepositoryName } from "./isRepositoryName";
export { isRepositoryEndpoint } from "./isRepositoryEndpoint";
export { buildQueryURL } from "./buildQueryURL";
export { createClient, Client } from "./client";

// Predicates API.
import { predicate } from "./predicate";
export { predicate };
/**
 * @deprecated Renamed to `predicate` (without an "s").
 */
// TODO: Remove in v3.
export const predicates = predicate;
/**
 * @deprecated Renamed to `predicate` (lowercase and without an "s").
 */
// TODO: Remove in v3.
export const Predicates = predicate;

// Custom errors used by Client.
export { PrismicError } from "./PrismicError";
export { ForbiddenError } from "./ForbiddenError";
export { ParsingError } from "./ParsingError";
export { NotFoundError } from "./NotFoundError";

// A collection of well-known cookie names shared between Prismic libraries and systems.
export * as cookie from "./cookie";

// General types used throughout the project. These are made public to allow users to better type their projects.
export type { CreateClient, ClientConfig } from "./client";
export type { QueryParams, BuildQueryURLArgs } from "./buildQueryURL";
export type {
	AbortSignalLike,
	FetchLike,
	HttpRequestLike,
	Ordering,
	RequestInitLike,
	ResponseLike,
	Route,
} from "./types";
