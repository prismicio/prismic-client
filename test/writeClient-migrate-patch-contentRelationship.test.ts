import { testMigrationFieldPatching } from "./__testutils__/testMigrationFieldPatching"

import type {
	ContentRelationshipField,
	InjectMigrationSpecificTypes,
	MigrationContentRelationship,
	RichTextField,
} from "../src"
import { LinkType, RichTextNodeType } from "../src"

testMigrationFieldPatching<
	MigrationContentRelationship | InjectMigrationSpecificTypes<RichTextField>
>("patches content relationship fields", {
	existing: ({ existingDocuments }) => existingDocuments[0],
	existingLongForm: ({ existingDocuments }) => {
		return {
			link_type: LinkType.Document,
			id: existingDocuments[0],
		}
	},
	existingLongFormWithText: ({ existingDocuments }) => {
		return {
			link_type: LinkType.Document,
			id: existingDocuments[0],
			text: "foo",
		}
	},
	otherCreate: ({ otherCreateDocument }) => otherCreateDocument,
	lazyExisting: ({ existingDocuments }) => {
		return () => existingDocuments[0]
	},
	lazyOtherCreate: ({ migration, otherCreateDocument }) => {
		return () =>
			migration.getByUID(
				otherCreateDocument.document.type,
				otherCreateDocument.document.uid!,
			)
	},
	lazyOtherCreateMissingID: ({ migration, otherCreateDocument }) => {
		return () => {
			const doc = migration.getByUID(
				otherCreateDocument.document.type,
				otherCreateDocument.document.uid!,
			)

			delete doc?.document.id

			return doc
		}
	},
	richTextLinkNode: ({ existingDocuments }) => [
		{
			type: RichTextNodeType.paragraph,
			text: "lorem",
			spans: [
				{ type: RichTextNodeType.strong, start: 0, end: 5 },
				{
					type: RichTextNodeType.hyperlink,
					start: 0,
					end: 5,
					data: existingDocuments[0],
				},
			],
		},
	],
})

testMigrationFieldPatching<ContentRelationshipField>(
	"patches content relationship fields (from Prismic)",
	{
		simple: ({ ctx, otherFromPrismicDocument }) => {
			const contentRelationship = ctx.mock.value.link({
				type: LinkType.Document,
			})
			// `migrationDocuments` contains documents from "another repository"
			contentRelationship.id =
				otherFromPrismicDocument.originalPrismicDocument!.id

			return contentRelationship
		},
		withText: ({ ctx, otherFromPrismicDocument }) => {
			const contentRelationship = ctx.mock.value.link({
				type: LinkType.Document,
			})
			// `migrationDocuments` contains documents from "another repository"
			contentRelationship.id =
				otherFromPrismicDocument.originalPrismicDocument!.id

			// TODO: Remove when link text PR is merged
			// @ts-expect-error - Future-proofing for link text
			contentRelationship.text = "foo"

			return contentRelationship
		},
		broken: () => {
			return {
				link_type: LinkType.Document,
				id: "id",
				type: "type",
				tags: [],
				lang: "lang",
				isBroken: true,
			}
		},
	},
	{ mode: "fromPrismic" },
)
