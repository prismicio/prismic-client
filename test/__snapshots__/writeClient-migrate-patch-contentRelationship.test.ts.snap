// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`patches content relationship fields (from Prismic) > broken > group 1`] = `
{
  "group": [
    {
      "field": {
        "id": "_____broken_____",
        "isBroken": true,
        "link_type": "Document",
      },
    },
  ],
}
`;

exports[`patches content relationship fields (from Prismic) > broken > shared slice 1`] = `
{
  "slices": [
    {
      "id": "9b9dd0f2c3f",
      "items": [
        {
          "field": {
            "id": "_____broken_____",
            "isBroken": true,
            "link_type": "Document",
          },
        },
      ],
      "primary": {
        "field": {
          "id": "_____broken_____",
          "isBroken": true,
          "link_type": "Document",
        },
        "group": [
          {
            "field": {
              "id": "_____broken_____",
              "isBroken": true,
              "link_type": "Document",
            },
          },
        ],
      },
      "slice_label": null,
      "slice_type": "nunc",
      "variation": "ullamcorper",
      "version": "8bfc905",
    },
  ],
}
`;

exports[`patches content relationship fields (from Prismic) > broken > slice 1`] = `
{
  "slices": [
    {
      "id": "5306297c5ed",
      "items": [
        {
          "field": {
            "id": "_____broken_____",
            "isBroken": true,
            "link_type": "Document",
          },
        },
      ],
      "primary": {
        "field": {
          "id": "_____broken_____",
          "isBroken": true,
          "link_type": "Document",
        },
      },
      "slice_label": "Vel",
      "slice_type": "hac_habitasse",
    },
  ],
}
`;

exports[`patches content relationship fields (from Prismic) > broken > static zone 1`] = `
{
  "field": {
    "id": "_____broken_____",
    "isBroken": true,
    "link_type": "Document",
  },
}
`;

exports[`patches content relationship fields (from Prismic) > simple > group 1`] = `
{
  "group": [
    {
      "field": {
        "id": "other.id-fromPrismic",
        "link_type": "Document",
      },
    },
  ],
}
`;

exports[`patches content relationship fields (from Prismic) > simple > shared slice 1`] = `
{
  "slices": [
    {
      "id": "9b9dd0f2c3f",
      "items": [
        {
          "field": {
            "id": "other.id-fromPrismic",
            "link_type": "Document",
          },
        },
      ],
      "primary": {
        "field": {
          "id": "other.id-fromPrismic",
          "link_type": "Document",
        },
        "group": [
          {
            "field": {
              "id": "other.id-fromPrismic",
              "link_type": "Document",
            },
          },
        ],
      },
      "slice_label": null,
      "slice_type": "nunc",
      "variation": "ullamcorper",
      "version": "8bfc905",
    },
  ],
}
`;

exports[`patches content relationship fields (from Prismic) > simple > slice 1`] = `
{
  "slices": [
    {
      "id": "5306297c5ed",
      "items": [
        {
          "field": {
            "id": "other.id-fromPrismic",
            "link_type": "Document",
          },
        },
      ],
      "primary": {
        "field": {
          "id": "other.id-fromPrismic",
          "link_type": "Document",
        },
      },
      "slice_label": "Vel",
      "slice_type": "hac_habitasse",
    },
  ],
}
`;

exports[`patches content relationship fields (from Prismic) > simple > static zone 1`] = `
{
  "field": {
    "id": "other.id-fromPrismic",
    "link_type": "Document",
  },
}
`;

exports[`patches content relationship fields (from Prismic) > withText > group 1`] = `
{
  "group": [
    {
      "field": {
        "id": "other.id-fromPrismic",
        "link_type": "Document",
        "text": "foo",
      },
    },
  ],
}
`;

exports[`patches content relationship fields (from Prismic) > withText > shared slice 1`] = `
{
  "slices": [
    {
      "id": "9b9dd0f2c3f",
      "items": [
        {
          "field": {
            "id": "other.id-fromPrismic",
            "link_type": "Document",
            "text": "foo",
          },
        },
      ],
      "primary": {
        "field": {
          "id": "other.id-fromPrismic",
          "link_type": "Document",
          "text": "foo",
        },
        "group": [
          {
            "field": {
              "id": "other.id-fromPrismic",
              "link_type": "Document",
              "text": "foo",
            },
          },
        ],
      },
      "slice_label": null,
      "slice_type": "nunc",
      "variation": "ullamcorper",
      "version": "8bfc905",
    },
  ],
}
`;

exports[`patches content relationship fields (from Prismic) > withText > slice 1`] = `
{
  "slices": [
    {
      "id": "5306297c5ed",
      "items": [
        {
          "field": {
            "id": "other.id-fromPrismic",
            "link_type": "Document",
            "text": "foo",
          },
        },
      ],
      "primary": {
        "field": {
          "id": "other.id-fromPrismic",
          "link_type": "Document",
          "text": "foo",
        },
      },
      "slice_label": "Vel",
      "slice_type": "hac_habitasse",
    },
  ],
}
`;

exports[`patches content relationship fields (from Prismic) > withText > static zone 1`] = `
{
  "field": {
    "id": "other.id-fromPrismic",
    "link_type": "Document",
    "text": "foo",
  },
}
`;

exports[`patches content relationship fields > existing > group 1`] = `
{
  "group": [
    {
      "field": {
        "id": "other.id-existing",
        "link_type": "Document",
      },
    },
  ],
}
`;

exports[`patches content relationship fields > existing > shared slice 1`] = `
{
  "slices": [
    {
      "id": "9b9dd0f2c3f",
      "items": [
        {
          "field": {
            "id": "other.id-existing",
            "link_type": "Document",
          },
        },
      ],
      "primary": {
        "field": {
          "id": "other.id-existing",
          "link_type": "Document",
        },
        "group": [
          {
            "field": {
              "id": "other.id-existing",
              "link_type": "Document",
            },
          },
        ],
      },
      "slice_label": null,
      "slice_type": "nunc",
      "variation": "ullamcorper",
      "version": "8bfc905",
    },
  ],
}
`;

exports[`patches content relationship fields > existing > slice 1`] = `
{
  "slices": [
    {
      "id": "5306297c5ed",
      "items": [
        {
          "field": {
            "id": "other.id-existing",
            "link_type": "Document",
          },
        },
      ],
      "primary": {
        "field": {
          "id": "other.id-existing",
          "link_type": "Document",
        },
      },
      "slice_label": "Vel",
      "slice_type": "hac_habitasse",
    },
  ],
}
`;

exports[`patches content relationship fields > existing > static zone 1`] = `
{
  "field": {
    "id": "other.id-existing",
    "link_type": "Document",
  },
}
`;

exports[`patches content relationship fields > existingLongForm > group 1`] = `
{
  "group": [
    {
      "field": {
        "id": "other.id-existing",
        "link_type": "Document",
      },
    },
  ],
}
`;

exports[`patches content relationship fields > existingLongForm > shared slice 1`] = `
{
  "slices": [
    {
      "id": "9b9dd0f2c3f",
      "items": [
        {
          "field": {
            "id": "other.id-existing",
            "link_type": "Document",
          },
        },
      ],
      "primary": {
        "field": {
          "id": "other.id-existing",
          "link_type": "Document",
        },
        "group": [
          {
            "field": {
              "id": "other.id-existing",
              "link_type": "Document",
            },
          },
        ],
      },
      "slice_label": null,
      "slice_type": "nunc",
      "variation": "ullamcorper",
      "version": "8bfc905",
    },
  ],
}
`;

exports[`patches content relationship fields > existingLongForm > slice 1`] = `
{
  "slices": [
    {
      "id": "5306297c5ed",
      "items": [
        {
          "field": {
            "id": "other.id-existing",
            "link_type": "Document",
          },
        },
      ],
      "primary": {
        "field": {
          "id": "other.id-existing",
          "link_type": "Document",
        },
      },
      "slice_label": "Vel",
      "slice_type": "hac_habitasse",
    },
  ],
}
`;

exports[`patches content relationship fields > existingLongForm > static zone 1`] = `
{
  "field": {
    "id": "other.id-existing",
    "link_type": "Document",
  },
}
`;

exports[`patches content relationship fields > existingLongFormWithText > group 1`] = `
{
  "group": [
    {
      "field": {
        "id": "other.id-existing",
        "link_type": "Document",
        "text": "foo",
      },
    },
  ],
}
`;

exports[`patches content relationship fields > existingLongFormWithText > shared slice 1`] = `
{
  "slices": [
    {
      "id": "9b9dd0f2c3f",
      "items": [
        {
          "field": {
            "id": "other.id-existing",
            "link_type": "Document",
            "text": "foo",
          },
        },
      ],
      "primary": {
        "field": {
          "id": "other.id-existing",
          "link_type": "Document",
          "text": "foo",
        },
        "group": [
          {
            "field": {
              "id": "other.id-existing",
              "link_type": "Document",
              "text": "foo",
            },
          },
        ],
      },
      "slice_label": null,
      "slice_type": "nunc",
      "variation": "ullamcorper",
      "version": "8bfc905",
    },
  ],
}
`;

exports[`patches content relationship fields > existingLongFormWithText > slice 1`] = `
{
  "slices": [
    {
      "id": "5306297c5ed",
      "items": [
        {
          "field": {
            "id": "other.id-existing",
            "link_type": "Document",
            "text": "foo",
          },
        },
      ],
      "primary": {
        "field": {
          "id": "other.id-existing",
          "link_type": "Document",
          "text": "foo",
        },
      },
      "slice_label": "Vel",
      "slice_type": "hac_habitasse",
    },
  ],
}
`;

exports[`patches content relationship fields > existingLongFormWithText > static zone 1`] = `
{
  "field": {
    "id": "other.id-existing",
    "link_type": "Document",
    "text": "foo",
  },
}
`;

exports[`patches content relationship fields > lazyExisting > group 1`] = `
{
  "group": [
    {
      "field": {
        "id": "other.id-existing",
        "link_type": "Document",
      },
    },
  ],
}
`;

exports[`patches content relationship fields > lazyExisting > shared slice 1`] = `
{
  "slices": [
    {
      "id": "9b9dd0f2c3f",
      "items": [
        {
          "field": {
            "id": "other.id-existing",
            "link_type": "Document",
          },
        },
      ],
      "primary": {
        "field": {
          "id": "other.id-existing",
          "link_type": "Document",
        },
        "group": [
          {
            "field": {
              "id": "other.id-existing",
              "link_type": "Document",
            },
          },
        ],
      },
      "slice_label": null,
      "slice_type": "nunc",
      "variation": "ullamcorper",
      "version": "8bfc905",
    },
  ],
}
`;

exports[`patches content relationship fields > lazyExisting > slice 1`] = `
{
  "slices": [
    {
      "id": "5306297c5ed",
      "items": [
        {
          "field": {
            "id": "other.id-existing",
            "link_type": "Document",
          },
        },
      ],
      "primary": {
        "field": {
          "id": "other.id-existing",
          "link_type": "Document",
        },
      },
      "slice_label": "Vel",
      "slice_type": "hac_habitasse",
    },
  ],
}
`;

exports[`patches content relationship fields > lazyExisting > static zone 1`] = `
{
  "field": {
    "id": "other.id-existing",
    "link_type": "Document",
  },
}
`;

exports[`patches content relationship fields > lazyOtherCreate > group 1`] = `
{
  "group": [
    {
      "field": {
        "id": "other.id-create",
        "link_type": "Document",
      },
    },
  ],
}
`;

exports[`patches content relationship fields > lazyOtherCreate > shared slice 1`] = `
{
  "slices": [
    {
      "id": "9b9dd0f2c3f",
      "items": [
        {
          "field": {
            "id": "other.id-create",
            "link_type": "Document",
          },
        },
      ],
      "primary": {
        "field": {
          "id": "other.id-create",
          "link_type": "Document",
        },
        "group": [
          {
            "field": {
              "id": "other.id-create",
              "link_type": "Document",
            },
          },
        ],
      },
      "slice_label": null,
      "slice_type": "nunc",
      "variation": "ullamcorper",
      "version": "8bfc905",
    },
  ],
}
`;

exports[`patches content relationship fields > lazyOtherCreate > slice 1`] = `
{
  "slices": [
    {
      "id": "5306297c5ed",
      "items": [
        {
          "field": {
            "id": "other.id-create",
            "link_type": "Document",
          },
        },
      ],
      "primary": {
        "field": {
          "id": "other.id-create",
          "link_type": "Document",
        },
      },
      "slice_label": "Vel",
      "slice_type": "hac_habitasse",
    },
  ],
}
`;

exports[`patches content relationship fields > lazyOtherCreate > static zone 1`] = `
{
  "field": {
    "id": "other.id-create",
    "link_type": "Document",
  },
}
`;

exports[`patches content relationship fields > lazyOtherCreateMissingID > group 1`] = `
{
  "group": [
    {
      "field": {
        "link_type": "Document",
      },
    },
  ],
}
`;

exports[`patches content relationship fields > lazyOtherCreateMissingID > shared slice 1`] = `
{
  "slices": [
    {
      "id": "9b9dd0f2c3f",
      "items": [
        {
          "field": {
            "link_type": "Document",
          },
        },
      ],
      "primary": {
        "field": {
          "link_type": "Document",
        },
        "group": [
          {
            "field": {
              "link_type": "Document",
            },
          },
        ],
      },
      "slice_label": null,
      "slice_type": "nunc",
      "variation": "ullamcorper",
      "version": "8bfc905",
    },
  ],
}
`;

exports[`patches content relationship fields > lazyOtherCreateMissingID > slice 1`] = `
{
  "slices": [
    {
      "id": "5306297c5ed",
      "items": [
        {
          "field": {
            "link_type": "Document",
          },
        },
      ],
      "primary": {
        "field": {
          "link_type": "Document",
        },
      },
      "slice_label": "Vel",
      "slice_type": "hac_habitasse",
    },
  ],
}
`;

exports[`patches content relationship fields > lazyOtherCreateMissingID > static zone 1`] = `
{
  "field": {
    "link_type": "Document",
  },
}
`;

exports[`patches content relationship fields > otherCreate > group 1`] = `
{
  "group": [
    {
      "field": {
        "id": "other.id-create",
        "link_type": "Document",
      },
    },
  ],
}
`;

exports[`patches content relationship fields > otherCreate > shared slice 1`] = `
{
  "slices": [
    {
      "id": "9b9dd0f2c3f",
      "items": [
        {
          "field": {
            "id": "other.id-create",
            "link_type": "Document",
          },
        },
      ],
      "primary": {
        "field": {
          "id": "other.id-create",
          "link_type": "Document",
        },
        "group": [
          {
            "field": {
              "id": "other.id-create",
              "link_type": "Document",
            },
          },
        ],
      },
      "slice_label": null,
      "slice_type": "nunc",
      "variation": "ullamcorper",
      "version": "8bfc905",
    },
  ],
}
`;

exports[`patches content relationship fields > otherCreate > slice 1`] = `
{
  "slices": [
    {
      "id": "5306297c5ed",
      "items": [
        {
          "field": {
            "id": "other.id-create",
            "link_type": "Document",
          },
        },
      ],
      "primary": {
        "field": {
          "id": "other.id-create",
          "link_type": "Document",
        },
      },
      "slice_label": "Vel",
      "slice_type": "hac_habitasse",
    },
  ],
}
`;

exports[`patches content relationship fields > otherCreate > static zone 1`] = `
{
  "field": {
    "id": "other.id-create",
    "link_type": "Document",
  },
}
`;

exports[`patches content relationship fields > richTextLinkNode > group 1`] = `
{
  "group": [
    {
      "field": [
        {
          "spans": [
            {
              "end": 5,
              "start": 0,
              "type": "strong",
            },
            {
              "data": {
                "id": "other.id-existing",
                "link_type": "Document",
              },
              "end": 5,
              "start": 0,
              "type": "hyperlink",
            },
          ],
          "text": "lorem",
          "type": "paragraph",
        },
      ],
    },
  ],
}
`;

exports[`patches content relationship fields > richTextLinkNode > shared slice 1`] = `
{
  "slices": [
    {
      "id": "9b9dd0f2c3f",
      "items": [
        {
          "field": [
            {
              "spans": [
                {
                  "end": 5,
                  "start": 0,
                  "type": "strong",
                },
                {
                  "data": {
                    "id": "other.id-existing",
                    "link_type": "Document",
                  },
                  "end": 5,
                  "start": 0,
                  "type": "hyperlink",
                },
              ],
              "text": "lorem",
              "type": "paragraph",
            },
          ],
        },
      ],
      "primary": {
        "field": [
          {
            "spans": [
              {
                "end": 5,
                "start": 0,
                "type": "strong",
              },
              {
                "data": {
                  "id": "other.id-existing",
                  "link_type": "Document",
                },
                "end": 5,
                "start": 0,
                "type": "hyperlink",
              },
            ],
            "text": "lorem",
            "type": "paragraph",
          },
        ],
        "group": [
          {
            "field": [
              {
                "spans": [
                  {
                    "end": 5,
                    "start": 0,
                    "type": "strong",
                  },
                  {
                    "data": {
                      "id": "other.id-existing",
                      "link_type": "Document",
                    },
                    "end": 5,
                    "start": 0,
                    "type": "hyperlink",
                  },
                ],
                "text": "lorem",
                "type": "paragraph",
              },
            ],
          },
        ],
      },
      "slice_label": null,
      "slice_type": "nunc",
      "variation": "ullamcorper",
      "version": "8bfc905",
    },
  ],
}
`;

exports[`patches content relationship fields > richTextLinkNode > slice 1`] = `
{
  "slices": [
    {
      "id": "5306297c5ed",
      "items": [
        {
          "field": [
            {
              "spans": [
                {
                  "end": 5,
                  "start": 0,
                  "type": "strong",
                },
                {
                  "data": {
                    "id": "other.id-existing",
                    "link_type": "Document",
                  },
                  "end": 5,
                  "start": 0,
                  "type": "hyperlink",
                },
              ],
              "text": "lorem",
              "type": "paragraph",
            },
          ],
        },
      ],
      "primary": {
        "field": [
          {
            "spans": [
              {
                "end": 5,
                "start": 0,
                "type": "strong",
              },
              {
                "data": {
                  "id": "other.id-existing",
                  "link_type": "Document",
                },
                "end": 5,
                "start": 0,
                "type": "hyperlink",
              },
            ],
            "text": "lorem",
            "type": "paragraph",
          },
        ],
      },
      "slice_label": "Vel",
      "slice_type": "hac_habitasse",
    },
  ],
}
`;

exports[`patches content relationship fields > richTextLinkNode > static zone 1`] = `
{
  "field": [
    {
      "spans": [
        {
          "end": 5,
          "start": 0,
          "type": "strong",
        },
        {
          "data": {
            "id": "other.id-existing",
            "link_type": "Document",
          },
          "end": 5,
          "start": 0,
          "type": "hyperlink",
        },
      ],
      "text": "lorem",
      "type": "paragraph",
    },
  ],
}
`;
